### ISSUE TITLE
Image with URL source that is unreachable leaves instance stuck in starting state on launch

### ISSUE THREAD

**Issue by rcgoodfellow** [oxidecomputer/omicron#1745]
If an image is created with the URL source type, but that URL is not reachable from the propolis zone launching the instance, the sled-agent instance-launch process fails and cleans up the propolis zone, but the instance remains stuck in a starting state.

**Comment by jmpesp** [comment]
It's surprising to me that the first part of this would happen - when an image is created with the URL source type, Nexus [performs a HEAD request to validate that the URL is reachable and that the HTTP server isn't returning an error for the path](https://github.com/oxidecomputer/omicron/blob/508906b5465aeb4a5e8dce4406fab0f708dbbd89/nexus/src/app/image.rs#L140-L155). If the URL is reachable from Nexus but not the Propolis zone, then that's definitely a bug.

What revs of Omicron and Propolis did you see this on? Could you attach the Nexus and Sled agent logs?

**Comment by rcgoodfellow** [comment]
The address that I was using was a local IPv4 address (e.g. I was running an HTTP server on the same machine as the virtual sled). This is the `192.168.1.5` address that shows up in the logs below. AFICT, there is no v4 plumbing between the propolis zones and the global zone? When I changed to using a v6 address, things worked. Maybe the Nexus zone has v4 plumbing, but Propolis zones do not?

This occurred on omicron rev ef58a07dfe165f190f8666d16d2d6b76beda9c7e. Propolis is the version that gets installed/used by omicron. I don't have this environment alive anymore. Happy to stand it up again and take a look at the propolis version.

I've attached some logs from a run where this happened. This is a slightly older rev of Nexus (prior to 92cf5cba7c51574a175244d30380f2f24d6a2e98). However, I did continue to observe this behavior after updating to ef58a07dfe165f190f8666d16d2d6b76beda9c7e. I can do another run from the latest Nexus and collect logs if that's helpful.

[nexus.log](https://github.com/oxidecomputer/omicron/files/9650088/nexus.log)
[sled-agent.log](https://github.com/oxidecomputer/omicron/files/9650089/sled-agent.log)

**Comment by jmpesp** [comment]
> Maybe the Nexus zone has v4 plumbing, but Propolis zones do not?

Yeah, this is true, as far as I can remember. The way to verify this is to race to run `ipadm` before the Propolis zone is torn down.

> I've attached some logs from a run where this happened. This is a slightly older rev of Nexus (prior to https://github.com/oxidecomputer/omicron/commit/92cf5cba7c51574a175244d30380f2f24d6a2e98). However, I did continue to observe this behavior after updating to https://github.com/oxidecomputer/omicron/commit/ef58a07dfe165f190f8666d16d2d6b76beda9c7e. I can do another run from the latest Nexus and collect logs if that's helpful.

Yes please, if those previously attached logs are for a rev that doesn't include 92cf5cba7c51574a175244d30380f2f24d6a2e98.

I was going to suggest updating to a rev that included 92cf5cba7c51574a175244d30380f2f24d6a2e98 before reading this comment, but it looks like you beat me to it :)

**Comment by rcgoodfellow** [comment]
Logs attached for from a post-https://github.com/oxidecomputer/omicron/commit/92cf5cba7c51574a175244d30380f2f24d6a2e98 rev. These logs contain two instance launches. The first is a successful one. For the second, I used an IPv4 image URL to get to the instance stuck in the starting state.

[nexus.log](https://github.com/oxidecomputer/omicron/files/9659807/nexus.log)
[sled-agent.log](https://github.com/oxidecomputer/omicron/files/9659808/sled-agent.log)

**Comment by davepacheco** [comment]
> If the URL is reachable from Nexus but not the Propolis zone, then that's definitely a bug.

Why is that?  If nothing else, there's a sort of time-of-check-time-of-use race here.  But isn't it possible that these two copmonents are (deliberately) on different networks?  I guess along those lines: if you create an image with an image source S on the public internet, is our assumption that your VM needs to be able to reach the public internet to download the image?

**Comment by jmpesp** [comment]
> Why is that?

It definitely could be a TOCTOU issue, yeah :) But the URL based image is only a stop-gap for development until we roll out the ability to upload an image straight to a Crucible region. With that, we won't have the addressability problem, and as a result won't have the TOCTOU problem. Nexus checking that it can reach the URL is to check for us misconfiguring things.

> Logs attached for from a post-92cf5cb rev.

This line from nexus.log:

```
[2022-09-27T21:37:43.682204368Z] ERROR: e6bff1ff-24fb-49dc-a54e-c6a350cd4d6c/ServerContext/8343 on oxz_nexus: saw Ok(false) from setting InstanceState::Failed after bad instance_put
```

where the `Ok(false)` means that `instance_update_runtime` did not update the row. The `execute_and_check` returned `UpdateStatus::NotUpdatedButExists`, which means `The row exists, but was not updated.`. I'm curious why the row was not updated, because I think this is only returned when all the filters on the `diesel::update` are matched correctly (because otherwise you'd see an error instead?). @davepacheco now that you're here - is this correct?

**Comment by davepacheco** [comment]
> I'm curious why the row was not updated, because I think this is only returned when all the filters on the diesel::update are matched correctly (because otherwise you'd see an error instead?). @davepacheco now that you're here - is this correct?

Paging this back in... I believe that return value means that that the row was found (that is, an Instance with the given instance_id was found), but the filters on the "update" part (the ones in `DataStore::instance_update_runtime()`) did not match the row.  It looks like this could happen if the Instance had since been deleted, or if the state_generation is older (or the same generation) as the one that's in the database, or there's a migration in progress to a different propolis.

**Comment by jmpesp** [comment]
> It looks like this could happen if the Instance had since been deleted, or if the state_generation is older (or the same generation) as the one that's in the database, or there's a migration in progress to a different propolis.

Hmm. We know that the instance was not deleted in parallel by Ry in this case (if so, please close this issue and open another one! :)), and that there isn't a migration occurring. [The Generation number being older shouldn't be possible, the code that is called when `instance_put` fails will increment the generation number before calling `instance_update_runtime`](https://github.com/oxidecomputer/omicron/blob/74f3ca89af11b0ce6d9f9bd4b5bdcbeb04d1ba3e/nexus/src/app/instance.rs#L657-L663). `sic_instance_ensure` is the last saga node being executed, so it's not a case where another saga node being unwound has deleted the instance record before we try to do this.

I don't have any idea why that `instance_update_runtime` didn't update the row.

**Comment by davepacheco** [comment]
Yes, the generation number that's used for the `instance_update_runtime()` call was just incremented, but it was incremented from the value in `db_instance`, which was fetched previously.  At the very least, it was in a separate transaction, so it's definitely _possible_ for something else to slip in and update the copy in the database between when we fetched `db_instance` and when we attempted to write the new value.  One candidate would be that if the Sled Agent somehow partially applied [this update](https://github.com/oxidecomputer/omicron/blob/74f3ca89af11b0ce6d9f9bd4b5bdcbeb04d1ba3e/nexus/src/app/instance.rs#L607-L616), it may have issued a request back to Nexus with the new runtime state (via `notify_instance_updated()`) and we may have received and processed that event.  If that were true, we'd have log entries for it.

I think the original intent of this code was that we would have "info" level log entries for every attempt to update the generation number (including whether it was successful or not) so that we could debug problems like this.  If there are code paths that don't do that, maybe we should update them, and then I'd think we could tell from the logs what happened.

**Comment by jmpesp** [comment]
(ignore the `Teach Nexus about the Pantry` messages, I pulled out the fix and made a separate PR)

**Comment by gjcolombo** [comment]
I encountered a variation of this just now. My local setup has a configuration problem (not an Omicron issue) that causes sled agent to fail to create instances if they have NICs attached. This leads to the following sequence of events:

1. Instance creation saga sends a start request to the sled via `instance_set_runtime`; the initial generation number for this instance's state is 1
2. `instance_set_runtime` calls sled agent's instance ensure endpoint to create the instance
3. Sled agent starts setting up the instance and reaches its `setup_propolis_locked` routine. This routine takes the original DB record, sets its state to Starting, bumps its generation to 2, and immediately kicks the updated record back up to Nexus: https://github.com/oxidecomputer/omicron/blob/f63cefae5f874692d307e66b270f234d86899469/sled-agent/src/instance.rs#L523-L543
4. The attempt to start the instance fails for unrelated reasons, so sled agent unwinds and returns an error to Nexus.
5. `instance_set_runtime` tries to handle the error by setting the instance's state to Failed and advancing the generation number. However, it does this using the original record it got as a parameter, not what's currently in the DB (line 661): https://github.com/oxidecomputer/omicron/blob/74f3ca89af11b0ce6d9f9bd4b5bdcbeb04d1ba3e/nexus/src/app/instance.rs#L618-L678
6. The update fails because `instance_set_runtime` is trying to write state generation 2, which sled agent already wrote in step 3.

Note that when this happens, you can't delete the instance, because we forbid attempts to delete an instance that's in the starting state. Stopping the instance might work, but only if the failure to create it was transient--if the problem recurs then #2517 will prevent you from unwedging the instance (since asking to stop it will attempt to start it again, causing the instance to land right back in the starting state).

ISTM we could improve this in a couple of ways:

1. Sled agent should broadcast that an instance failed if it failed to start it.
2. Nexus needs an instance heartbeat to keep instances from getting stuck in an inoperable state when they're completely lost. (If, in the example above, the sled agent had panicked after step 3, the "I failed to start this instance" update would never have been sent.)
